bin/codemods/src/single-tree-rendering.js

```diff
@@ -0,0 +1,107 @@
+/** @format */
+
+const config = require( './config' );
+
+export default function transformer( file, api ) {
+	const j = api.jscodeshift;
+	const root = j( file.source );
+
+	function addNextMiddleware( path ) {
+		if ( path.value.params.length !== 1 ) {
+			// More than just a context arg, possibly not a middleware
+			return path.value;
+		}
+		let ret = path.value;
+		ret.params = [ ...ret.params, j.identifier( 'next' ) ];
+		ret.body = j.blockStatement( [
+			...path.value.body.body,
+			j.expressionStatement( j.callExpression( j.identifier( 'next' ), [] ) ),
+		] );
+
+		return ret;
+	}
+
+	function getTarget( arg ) {
+		if ( arg.type === 'Literal' ) {
+			return arg.value;
+		}
+		if ( arg.type === 'CallExpression' ) {
+			// More checks?
+			return arg.arguments[ 0 ].value;
+		}
+	}
+
+	function transformRenderWithReduxStore( path ) {
+		if ( path.value.params.length !== 2 ) {
+			// More than just context and next args, possibly not a middleware
+			return path.value;
+		}
+		return j( path )
+			.find( j.CallExpression, {
+				callee: {
+					name: 'renderWithReduxStore',
+				},
+			} )
+			.replaceWith( p => {
+				const contextArg = path.value.params[ 0 ];
+				const target = getTarget( p.value.arguments[ 1 ] );
+				return j.assignmentExpression(
+					'=',
+					j.memberExpression( contextArg, j.identifier( target ) ),
+					p.value.arguments[ 0 ]
+				);
+			} );
+	}
+
+	root
+		.find( j.CallExpression, {
+			callee: {
+				name: 'renderWithReduxStore',
+			},
+		} )
+		.closestScope()
+		.replaceWith( addNextMiddleware )
+		.forEach( transformRenderWithReduxStore );
+
+	// Remove `renderWithReduxStore` from `import { a, renderWithReduxStore, b } from 'lib/react-helpers'`
+	root
+		.find( j.ImportSpecifier, {
+			local: {
+				name: 'renderWithReduxStore',
+			},
+		} )
+		.remove();
+
+	// Remove empty `import 'lib/react-helpers'`
+	root
+		.find( j.ImportDeclaration, {
+			source: {
+				value: 'lib/react-helpers',
+			},
+		} )
+		.filter( p => ! p.value.specifiers.length )
+		.remove();
+
+	// Add makeLayout and clientRender middlewares to route definitions
+	const routeDefs = root
+		.find( j.CallExpression, {
+			callee: {
+				name: 'page',
+			},
+		} )
+		.filter( p => p.value.arguments.length > 1 && p.value.arguments[ 0 ].value !== '*' )
+		.forEach( p => {
+			p.value.arguments.push( j.identifier( 'makeLayout' ) );
+			p.value.arguments.push( j.identifier( 'clientRender' ) );
+		} );
+
+	if ( routeDefs.nodes().length ) {
+		// This adds a newline above the import, meh. https://github.com/benjamn/recast/issues/371
+		root
+			.find( j.ImportDeclaration )
+			.at( -1 )
+			.insertAfter( "import { makeLayout, render as clientRender } from 'controller';" );
```

Comment:
BTW since there was an issue of empty lines appearing above this new line, I'll want to check if it's because we're inserting a string. Maybe it'll behave differently if we insert pure `j.ImportDeclaration` object.

Running Prettify after the codemod gets rid of multiple empty lines, but still leaves one empty line. Not sure if we need to care about that, tho...

Reply:
> Similarly like above with `makeLayout`/`clientRender` identifiers, this isn't checking if `import 'controller'` already exists so it's possible to end up with this same line multiple times.

Yeah. Oh well, that's one of the instances where I think iterating might be fine -- we merge this PR, run it on some directory (or the entire codebase), see what's not transformed correctly, adapt the codemod accordingly, and repeat 😄 

> I was basically running this codemod to files which had this exact same line (`./client/me/help/index.js` for example) but I'm not sure if we need to worry about that: we would just not run the codemod on those files.

Hmm, the point of a codemod is to be able to run it at a large scale -- we wouldn't want to manually compile lists of files we want to run it on.

> I'd imagine that there would be situations where something else (like `navigation`) is already being imported from that `controller`. I'll have to check if this kind of situations actually exist.

That's a different file tho -- `makeLayout`  and `clientRender` live in `client/`, whereas `navigation` and `siteSelection` are in `client/my-sites/`.

Reply:
> BTW since there was an issue of empty lines appearing above this new line, I'll want to check if it's because we're inserting a string. Maybe it'll behave differently if we insert pure `j.ImportDeclaration` object? Like so: `insertAfter( j.importDeclaration( ... ) )`

Worth a try I guess. Wanna try it out? (Could be deemed an improvement we tackle separately after merging this PR, too.)

> Running Prettify right after the codemod does get rid of multiple empty lines, but still leaves one empty line. Not sure if we need to care about that, tho...

Ideally we get rid of it, but we shouldn't get blocked by it.

Reply:
>>BTW since there was an issue of empty lines appearing above this new line, I'll want to check if it's because we're inserting a string. Maybe it'll behave differently if we insert pure j.ImportDeclaration object? Like so: insertAfter( j.importDeclaration( ... ) )

>Worth a try I guess. Wanna try it out? (Could be deemed an improvement we tackle separately after merging this PR, too.)

I actually did try this last night, but mostly just to learn first hand how to modify ASTs. This is so many lines vs just one line you wrote and end result is exactly the same. Honestly I don't think it's worth changing, just makes it harder to read:
```js
// Generates `makeLayout` for importing
const makeLayout = j.importSpecifier( j.identifier( 'makeLayout' ) );

// Generates `render as clientRender` for importing
let renderAsClientRender = j.importSpecifier(
  j.identifier( 'render' ) // Adds "imported" identifier
);
renderAsClientRender.local = j.identifier( 'clientRender' ); // Adds "local" identifier

root
  .find( j.ImportDeclaration )
  .at( -1 )
  .insertAfter( () => {
    // Generates `import { makeLayout, render as clientRender } from 'controller';`
    return j.importDeclaration(
      [
        makeLayout,
        renderAsClientRender
      ],
      j.literal( 'controller' )
    );
  } );
```

>>Running Prettify right after the codemod does get rid of multiple empty lines, but still leaves one empty line. Not sure if we need to care about that, tho...
>Ideally we get rid of it, but we shouldn't get blocked by it.

I actually saw a simple regexp fix for this elsewhere, gonna give it a try:
https://github.com/Automattic/wp-calypso/blob/fix/codemod-export-config/bin/codemods/src/sort-imports.js#L30-L33

Reply:
> I actually did try this last night, but mostly just to learn first hand how to modify ASTs. This is so many lines vs just one line you wrote and end result is exactly the same. Honestly I don't think it's worth changing, just makes it harder to read: [...]

Well the key part here is that the _end result is exactly the same_. Not worth making the change then! 😄 

> I actually saw a simple regexp fix for this elsewhere, gonna give it a try:
> https://github.com/Automattic/wp-calypso/blob/master/bin/codemods/src/sort-imports.js#L30-L33

👍